Firestore/core/src/firebase/firestore/local/index_free_query_engine.h

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_INDEX_FREE_QUERY_ENGINE_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_INDEX_FREE_QUERY_ENGINE_H_
+
+#include "Firestore/core/src/firebase/firestore/local/query_engine.h"
+
+#include "Firestore/core/src/firebase/firestore/model/document_key_set.h"
+#include "Firestore/core/src/firebase/firestore/model/document_map.h"
+
+namespace firebase {
+namespace firestore {
+
+namespace model {
+
+class DocumentSet;
+
+}  // namespace model
+
+namespace local {
+
+/**
+ * A query engine that takes advantage of the target document mapping in the
+ * QueryCache. The IndexFreeQueryEngine optimizes query execution by only
+ * reading the documents that previously matched a query plus any documents that
+ * were edited after the query was last listened to.
+ *
+ * There are some cases where Index-Free queries are not guaranteed to
+ * produce the same results as full collection scans. In these cases, the
+ * IndexFreeQueryEngine falls back to full query processing. These cases are:
+ *
+ * - Limit queries where a document that matched the query previously no
+ *   longer matches the query.
+ * - Limit queries where a document edit may cause the document to sort below
+ *   another document that is in the local cache.
+ * - Queries that have never been CURRENT or free of Limbo documents.
+ */
+class IndexFreeQueryEngine : public QueryEngine {
+ public:
+  void SetLocalDocumentsView(LocalDocumentsView* local_documents) override {
+    local_documents_view_ = local_documents;
+  }
+
+  model::DocumentMap GetDocumentsMatchingQuery(
```

Comment:
Optional: I'm not sure this function should be `const`. IIUC, it might modify the local database, which doesn't feel like a `const` operation. Note that you can call non-const member functions on `local_documents_view_` inside a `const` member function, but I'd consider this more of a quirk of the language that `const` isn't "transitive" in this case.

Feel free to postpone/ignore if changing the base class interface is too much trouble.

Reply:
I moved this discussion to Chat. I am mostly surprised that other `get()` functions are not marked const, including the ones in LocalDocumentsView. It's not a huge change to drop the const (the QueryEngine interface is also new), but I want to make sure we are all on the same page.

Reply:
Update: I dropped `const` for this method and for ExecuteFullCollectionScan.